addons/product/tests/test_variants.py

```diff
@@ -922,6 +922,98 @@ def test_name_search_dynamic_attributes(self):
         name_searched = self.env['product.template'].name_search(name='cima')
         self.assertIn(template.id, [ng[0] for ng in name_searched])
 
+    def test_dynamic_attributes_archiving(self):
+        Product = self.env['product.product']
+        ProductAttribute = self.env['product.attribute']
+        ProductAttributeValue = self.env['product.attribute.value']
+
+        # Creating attributes
+        pa_color = ProductAttribute.create({'sequence': 1, 'name': 'color', 'create_variant': 'dynamic'})
+        color_values = ProductAttributeValue.create([{
+            'name': n,
+            'sequence': i,
+            'attribute_id': pa_color.id,
+        } for i, n in enumerate(['white', 'black'])])
+        pav_color_white = color_values[0]
+        pav_color_black = color_values[1]
+
+        pa_size = ProductAttribute.create({'sequence': 2, 'name': 'size', 'create_variant': 'dynamic'})
+        size_values = ProductAttributeValue.create([{
+            'name': n,
+            'sequence': i,
+            'attribute_id': pa_size.id,
+        } for i, n in enumerate(['s', 'm'])])
+        pav_size_s = size_values[0]
+        pav_size_m = size_values[1]
+
+        # Define a template with only color attribute & white value
+        template = self.env['product.template'].create({
+            'name': 'test product',
+            'attribute_line_ids': [(0, 0, {
+                'attribute_id': pa_color.id,
+                'value_ids': [(6, 0, [pav_color_white.id])],
+            })],
+        })
+
+        # Patch unlink method to force archiving instead deleting
+        def unlink(self):
+            self.active = False
+        Product._patch_method('unlink', unlink)
+
+        # Create a variant (because of dynamic attribute)
+        ptav_white = self.env['product.template.attribute.value'].search([
+            ('attribute_line_id', '=', template.attribute_line_ids.id),
+            ('product_attribute_value_id', '=', pav_color_white.id)
+        ])
+
+        product_white = template._create_product_variant(ptav_white)
+
+        # Adding a new value to an existing attribute should not archive the variant
+        template.write({
+            'attribute_line_ids': [(1, template.attribute_line_ids[0].id, {
+                'attribute_id': pa_color.id,
+                'value_ids': [(4, pav_color_black.id, False)],
+            })]
+        })
+
+        self.assertTrue(product_white.active)
+
+        # Removing an attribute value should archive the product using it
+        template.write({
+            'attribute_line_ids': [(1, template.attribute_line_ids[0].id, {
+                'value_ids': [(3, pav_color_white.id, 0)],
+            })]
+        })
+
+        self.assertFalse(product_white.active)
+
+        # Creating a product with the same attributes for testing duplicates
+        product_white_duplicate = Product.with_context(create_product_product=True).create({
```

Comment:
I don't think you need to set this context, it's already done in create of `product.product`.